### PR TITLE
Make the forward pass `torch.compile` compatible

### PR DESCRIPTION
**Pull request by MekkCyber**
This PR updates the flow of replace_kernel_forward_from_hub to ensure the appropriate forward pass is selected before model compilation. It introduces a new function `kernelize` that can be called to replace the forward before the model is compiled. It can also be called after registering new kernels for more flexibility

### CODE REVIEW

File: src/kernels/layer.py [hunk 1]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
 
 
-def replace_kernel_forward_from_hub(cls, layer_name: str, *, use_fallback: bool = True):
-    """
-    Replace the forward function of a layer using a layer from the kernel hub.
-    This function monkeypatches a layer, replacing the `forward` method
-    of the layer with that of a layer from the hub. The replacement is done
-    when a layer matching `layer_name` and device type is registered through
-    `register_layer_mapping`. The device type is inferred from the first
-    argument to `forward`.
```
Comment (danieldk) [inline src/kernels/layer.py:129]: The docstring got removed.

File: src/kernels/layer.py [hunk 2]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
+def replace_kernel_forward_from_hub(
+    cls,
+    layer_name: str,
+    *,
+    device: Device = Device(type="cuda"),
+    needs_backward: bool = False,
+    use_fallback: bool = True,
+):
+    cls.kernel_layer_name = layer_name
+    kernelize(cls, traverse_graph=False, device=device, needs_backward=needs_backward, use_fallback=use_fallback)
```
Comment (danieldk) [inline src/kernels/layer.py]: I think we should remove `kernelize` here and remove the `needs_backward` and `device` parameters from `replace_kernel_forward_from_hub`. This function and the decorator should just set up a class <-> name mapping and not mutate the classes (we would also lose the fallbacks).

File: src/kernels/layer.py [hunk 3]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
 
-    fallback_forward = cls.forward
 
-    cached_layer: Dict[LayerRepository, nn.Module] = {}
+def kernelize(
+    module_,
+    traverse_graph: bool = True,
+    device: Device = Device(type="cuda"),
+    needs_backward: bool = False,
```
Comment (danieldk) [inline src/kernels/layer.py]: We can get this from `module_` itself by checking whether it's in eval or training mode.

File: src/kernels/layer.py [hunk 4]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
+    use_fallback: bool = True,
+):
+    cls.kernel_layer_name = layer_name
+    kernelize(cls, traverse_graph=False, device=device, needs_backward=needs_backward, use_fallback=use_fallback)
 
-    fallback_forward = cls.forward
 
-    cached_layer: Dict[LayerRepository, nn.Module] = {}
+def kernelize(
+    module_,
```
Comment (danieldk) [inline src/kernels/layer.py]:
```suggestion
    module: nn.Module,
```

File: src/kernels/layer.py [hunk 5]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
-                raise ValueError(
-                    f"No layer mapping for `{layer_name}` with device type `{device.type}`"
+    # We assume only one Module per repository.
+    cached_layer: Dict[LayerRepository, nn.Module] = {}
+    is_compiling = _is_torchdynamo_compiling()
+    # If we don't traverse the graph, we stop after the first module
+    modules_list = [("", module_)] if not traverse_graph else module_.named_modules()
+
+    for _, module in modules_list:
+        if hasattr(module, "kernel_layer_name") or not traverse_graph:
```
Comment (danieldk) [inline src/kernels/layer.py]: What is the ` or not traverse_graph` condition for? I the layer has no kernel layer name, there is nothing to do, right?

File: src/kernels/layer.py [hunk 6]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
-    """
+def replace_kernel_forward_from_hub(
+    cls,
+    layer_name: str,
+    *,
+    device: Device = Device(type="cuda"),
+    needs_backward: bool = False,
+    use_fallback: bool = True,
+):
+    cls.kernel_layer_name = layer_name
```
Comment (danieldk) [inline src/kernels/layer.py:145]: Maybe we should also store the fallback forward? Then we would really have the proper fallback even if someone runs `kernelize` multiple times.

File: src/kernels/layer.py [hunk 7]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
 
-    cls.forward = forward
+            # Switch to fallback when the layer does not support
+            # compilation/compile when needed.
+            needs_fallback = needs_backward and not getattr(layer, "has_backward", True)
+            needs_fallback |= is_compiling and not getattr(layer, "can_torch_compile", False)
+            if needs_fallback:
+                module.forward = fallback_forward
+                continue
+            module.forward = layer.forward
```
Comment (danieldk) [inline src/kernels/layer.py]:
```suggestion
            if needs_fallback:
                module.forward = fallback_forward
            else:
                module.forward = layer.forward
```

File: src/kernels/layer.py [hunk 8]
```diff
@@ -119,96 +111,118 @@ def register_kernel_mapping(
...
 
 
-def use_kernel_forward_from_hub(layer_name: str, *, use_fallback: bool = True):
+def use_kernel_forward_from_hub(
+    layer_name: str, *, device: Device = Device(type="cuda"), needs_backward: bool = False, use_fallback: bool = True
```
Comment (danieldk) [inline src/kernels/layer.py]: See above.

File: src/kernels/layer.py (hunk 5 above)
Comment (MekkCyber) [reply src/kernels/layer.py]: My idea was that we can make `kernelize` handle two cases, swap the forwards during the import, and give the possibility to swap even after the import, if some new kernels are added if `kernel_layer_name` is set, in the first case, we don't traverse the model graph, we only do that in the second case where we loop over the modules

File: src/kernels/layer.py (hunk 3 above)
Comment (MekkCyber) [reply src/kernels/layer.py]: Nice catch

File: src/kernels/layer.py (hunk 2 above)
Comment (MekkCyber) [reply src/kernels/layer.py]: but where would we call `kernelize` then ? And it's not really efficient to loop after the model is already imported I think

Comment (danieldk) [reply src/kernels/layer.py]: One issue that I see with this is that previously, you could simply disable kernels at no cost by e.g. using an empty mapping in a context manager. Now disabling it (if not done using the env var) entails rewriting the whole graph, whereas I think for the time being _not_ using kernels should be the simple/cheap case.

There was also a discussion on Slack that using kernels should probably be opt-in for the time being. One reason being that kernels tend to introduce more non-determinism, which is not nice for tests, etc. Explicitly calling `kernelize` facilitates that. It's pretty similar to `torch.compile` being opt-in. If we return the model itself in `kernelize`, users could even do

```
compiled = torch.compile(kernelize(model, ...), ...)
```

and have an optimized model.

Comment (MekkCyber) [reply src/kernels/layer.py]: To make avoiding kernels the simplest option, we could add a new argument to `kernelize` defaulting to `True` that explicitly disables the kernels. What do you think?

As for rewriting the whole graph, I don’t see it as an issue. I’d prefer to keep the forward pass fast by avoiding any kernel-switching logic there. Disabling kernels once via a call to `kernelize` seems much better than checking conditions repeatedly during each forward pass, added to that the compile compatibility concerns.

File: src/kernels/layer.py [hunk 9]
```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
...
+                # Switch to fallback when the layer does not support:
+                # compilation/compile when needed.
+                # backward when needed
+                needs_fallback = needs_backward and not getattr(layer, "has_backward", True)
+                needs_fallback |= is_compiling and not getattr(layer, "can_torch_compile", False)
+                if needs_fallback:
+                    module_class.forward = fallback_forward
+                    continue
+                module_class.forward = layer.forward
+                continue
```
Comment (danieldk) [inline src/kernels/layer.py]:
```suggestion
                if needs_fallback:
                    module_class.forward = fallback_forward
                else:
                    module_class.forward = layer.forward
                continue
```

File: src/kernels/layer.py [hunk 10]
```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
...
+        cls: The layer class to decorate
+        layer_name: The name of the layer to use for kernel lookup
     """
+    cls.kernel_layer_name = layer_name
+    cls.fallback_forward = cls.forward
 
-    fallback_forward = cls.forward
 
-    cached_layer: Dict[LayerRepository, nn.Module] = {}
+cached_layer: ContextVar[Dict[LayerRepository, nn.Module]] = ContextVar("cached_layer", default={})
```
Comment (danieldk) [inline src/kernels/layer.py]: Since this is now a global, I'd put it at the top of the file and name it `_CACHED_LAYER`.

We don't really use this as a context var, so I think this could just be a regular `Dict`?

File: src/kernels/layer.py [hunk 11]
```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
...
 
-    def forward(self, x, *args, **kwargs):
-        if _DISABLE_KERNEL_MAPPING:
-            return fallback_forward(self, x, *args, **kwargs)
+def kernelize(
+    model: nn.Module,
+    device: Device = Device(type="cuda"),
```
Comment (danieldk) [inline src/kernels/layer.py]: I think it would be nicer if this was:

```suggestion
    device: Optional[Device] = None,
```

Behavior: if the user passes a device type, use it. If the user does not pass a device type, get it from a (probably the first, since individual layers may not have parameters) parameter of the model. Similarly, I would add an extra argument:

```
    needs_backward: Optional[bool] = None,
```

If the user sets a bool, then do as they wish. Otherwise, use the logic that is already there below. This has the benefit that if a model is in `eval()`, the user could still kernelize it with kernels that support backward in case they want to move to `train()` later.

Then, with these new options, I would make everything after `model` kwarg-only. Positional arguments lead to a lot of bugs/failures.

File: src/kernels/layer.py [hunk 12]
```diff
@@ -8,8 +8,7 @@
 
 from .utils import get_kernel
 
-if TYPE_CHECKING:
-    from torch import nn
+from torch import nn
```
Comment (danieldk) [inline src/kernels/layer.py]: Needs to be conditional, `torch` is an optional dependency.

**Review by danieldk — COMMENTED** [review]
Looks great! Added some comments to polish it up.

File: src/kernels/layer.py [hunk 13]
```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
...
-    of the layer with that of a layer from the hub. The replacement is done
-    when a layer matching `layer_name` and device type is registered through
-    `register_layer_mapping`. The device type is inferred from the first
-    argument to `forward`.
+    Decorator that prepares a layer class to use a kernel from the Hugging Face Hub.
+    
+    This decorator stores the layer name and original forward method, which will be used
+    by the kernelize function to replace the forward implementation with the appropriate
+    kernel from the hub.
+
```
Comment (ArthurZucker) [inline src/kernels/layer.py]: let's explain why we don't replace on the fly (cf compile)

File: src/kernels/layer.py [hunk 14]
```diff
@@ -119,96 +110,125 @@ def register_kernel_mapping(
...
-        repo = kernel.get(Device(type=device.type))
-        if repo is None:
-            if not use_fallback:
-                raise ValueError(
-                    f"No layer mapping for `{layer_name}` with device type `{device.type}`"
+    Returns:
+        The kernelized model
+    """
+    is_compiling = _is_torchdynamo_compiling()
+    needs_backward = model.training
```
Comment (ArthurZucker) [inline src/kernels/layer.py]: does this mean we cannot adapt to the user changing `model.training` . 
Would be nice if we can register two forwards, and switch between them on the fly!

**Review by ArthurZucker — COMMENTED** [review]
Nice!

File: src/kernels/layer.py (hunk 13 above)
Comment (danieldk) [reply src/kernels/layer.py]: I added a FAQ with this question in it.

File: src/kernels/layer.py (hunk 14 above)
Comment (danieldk) [reply src/kernels/layer.py]: Yeah, that would be nice! I think to support this we would have to override `Module.train` and switch forwards based on the value that is passed. I think it makes sense that I do this in a PR after this, since this one is already getting pretty big.

Then we'll remove the `needs_backward` argument again in the next PR (which should be fine as long we don't make a release).

File: docs/layers.md [hunk 15]
```diff
@@ -23,33 +23,98 @@ class SiluAndMul(nn.Module):
...
+A model will not use Hub kernels by default, even if it contains extensible
+layers. To enable the use of Hub kernels in the model, it needs to be
+'kernelized' using the `kernelize` function. This function traverses the
+model graph and replaces the `forward` methods of extensible layers for which
+Hub kernels are registered. Kernelize can be used as follows:
+
+```python
+model = MyModel(...)
+model = kernelize(model)
+```
```
Comment (LysandreJik) [inline docs/layers.md:57]: I like that opt-in, that's clean

File: docs/layers.md [hunk 16]
```diff
@@ -23,33 +23,98 @@ class SiluAndMul(nn.Module):
...
 
-The `register_kernel_mapping` call maps the name `SiluAndMul` to actual
-hub kernels. See the [Registering a hub kernel for a layer](#registering-a-hub-kernel-for-a-layer)
-section for more information.
-
 **Warning:** we strongly recommend using layers with a decorator, since
 it signifies that the maintainer intends to keep the `forward` signature
 compatible with layers from the hub.
 
+## Kernelizing a model
```
Comment (LysandreJik) [inline docs/layers.md:46]: Very clear doc

**Review by LysandreJik — APPROVED** [review]
Very nice! Love the new additions and the clear doc, thanks for spending time on this
